### PR TITLE
🏗✨ Add a test for sourcemaps generation during minified compilation

### LINKED ISSUE #27681: Add Sourcemap "golden" tests
## Describe the new feature or change to an existing feature you'd like to see

Unfortunately, we're getting frequent breaks in our Sourcemap pipeline. Either the source locations are incorrect because we forget to propagate, or a part of the pipeline destroys the information.

@choumx suggested maybe we could create golden tests for our sourcemap pipeline. Meaning, we'd use our latest code in `build-system/` to compile a known commit of AMP (eg, `c106206e169d3ce14a83faaa0a29b92e6dd00c76`). That way, if any change in our build process causes a change in the sourcemaps, we'll be alerted.

### PR DESCRIPTION
**Pull request by rsimha**
**Background:**

During minified compilation, we generate a sourcemap file called `v0.js.map` in addition to the runtime file `v0.js`. Sourcemap generation [frequently breaks](https://github.com/ampproject/amphtml/pulls?utf8=%E2%9C%93&q=is%3Apr+is%3Aclosed+sourcemaps++sort%3Aupdated-desc) due to untested compilation toolchain changes. See #27681. 

**PR highlights:**
- Adds a task called `gulp check-sourcemaps`
- Compiles the runtime with sourcemaps (`gulp dist --core_runtime_only --full_sourcemaps`)
- Verifies the the `sourceRoot` field exists, and is of the correct format
- Verifies that the `sources` array exists, and all paths in it point to valid files
- Verifies that the `mappings` object exists, decodes it, and verifies that a mapping exists for a known sentinel piece of code

**Usage:**
```sh
# Build core runtime and check generated sourcemaps
gulp check-sourcemaps

# Check sourcemaps from previous invocation of gulp dist --full_sourcemaps
gulp check-sourcemaps --nobuild
```

**Passing example:**

image

**Failing example (incorrect `sourceRoot`: wrong sourcemap domain):**

image

**Failing example (incorrect `sources`: revert fix in #27665):**

image

**Failing example (incorrect `mappings`: modify sentinel piece of code):**

image

Fixes #27681

### CODE REVIEW

**Comment by jridgewell** [comment]
Looking at the [source code](https://github.com/Anthonyzou/Sourcemap-Unpack/blob/master/unmap.js), `source-map-unpacker` doesn't actually  regenerate the inputs from the sourcemap. It's just printing the included `sourcesContent` to a file. We need actual tests of the sourcema's mappings.

**Comment by rsimha** [comment]
> Looking at the [source code](https://github.com/Anthonyzou/Sourcemap-Unpack/blob/master/unmap.js) [](/Anthonyzou/Sourcemap-Unpack/blob/HEAD@{2020-04-16T02:37:44Z}/unmap.js), `source-map-unpacker` doesn't actually regenerate the inputs from the sourcemap. It's just printing the included `sourcesContent` to a file.

Pretty sure this is not true. I made a known change to the source code in `v0.js.map` and it showed up as a diff. From [this line](https://github.com/Anthonyzou/Sourcemap-Unpack/blob/master/unmap.js#L37), I'm pretty certain `js.sourcesContent[idx]` is coming from the map file, not from the original source.

**Comment by rsimha** [comment]
@jridgewell I've added a step that compares ASTs of `src/amp.js` and `dist/extracted/src/amp.js`. There are diffs which you will see in [these logs](https://travis-ci.org/github/ampproject/amphtml/jobs/675578790).

**Comment by rsimha** [comment]
As discussed offline with @jridgewell, we're still in search of an effective way to extract the mappings from `v0.js.map` and ensure their correctness (via say, a sentinel value). Until then, there's still some value in making sure the sourcemap file is generated, and the correct sourcemap URL is present.

I've modified the scope of this PR to add the `gulp check-sourcemaps` task and verify the sourcemaps URL. The rest will be done in a follow up PR when time permits.

**Comment by rsimha** [comment]
@choumx @jridgewell All changes made. This is now ready for review and merge.

File: build-system/tasks/check-sourcemaps.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    log(
+      red('ERROR:'),
+      'Could not find',
+      cyan('sourceRoot'),
+      'in',
+      cyan(v0JsMap)
+    );
+    throwError('Error in finding sourcemap URL');
+  }
+  if (!v0JsMapJson.sourceRoot.match(sourcemapUrlMatcher)) {
```
Comment (choumx) [inline build-system/tasks/check-sourcemaps.js:69]: Would this have caught one of the previous source map bugs? My read is "probably not" looking at the regex: `const sourcemapUrlMatcher = 'https://raw.githubusercontent.com/ampproject/amphtml/\\d{13}/'`.

The marginal CI cost appears to be non-zero, so I'd say let's ship if I'm wrong about the above or we can reduce the cost to zero. :)

Comment (rsimha) [reply build-system/tasks/check-sourcemaps.js:69]: @jridgewell was the previous breakage in the URL format that's being tested here? Or was it something else?

Comment (jridgewell) [reply build-system/tasks/check-sourcemaps.js:69]: The previous breakage had `https://raw.githubusercontent.com/ampproject/amphtml/123/src/service/src/service/foo.js` (duplicate `src/service/src/service/`, which wouldn't be caught by this.

Comment (rsimha) [reply build-system/tasks/check-sourcemaps.js:69]: Do either of you know which field contains that path? If that too is non-trivial, I'm okay with abandoning the PR based on this thread.

Comment (jridgewell) [reply build-system/tasks/check-sourcemaps.js:69]: `sources` array had these paths.

Comment (rsimha) [reply build-system/tasks/check-sourcemaps.js]: Gotcha. Added a check to make sure all paths in the `sources` array are valid.

**Comment by rsimha** [comment]
Added a new check for all the paths in the `sources` array.

**Comment by rsimha** [comment]
Bump. Can this be merged?

**Comment by rsimha** [comment]
Added yet another check that implements https://github.com/ampproject/amphtml/issues/27681#issuecomment-613712669 and inspects the `mappings` field for a known sentinel value.

**Pull request merged**